jpyinterpreter/src/main/java/org/optaplanner/jpyinterpreter/PythonCompiledFunction.java

```diff
@@ -200,32 +203,43 @@ public BiFunction<PythonLikeTuple, PythonLikeDict, ArgumentSpec<PythonLikeObject
             for (; variableIndex < co_posonlyargcount; variableIndex++) {
                 if (variableIndex >= defaultPositionalStartIndex) {
-                    out = out.addPositionalOnlyArgument(co_varnames.get(variableIndex), PythonLikeObject.class,
+                    out = out.addPositionalOnlyArgument(co_varnames.get(variableIndex),
+                            (Class) parameterTypeList.get(variableIndex)
+                                    .getJavaClassOrDefault(PythonLikeObject.class),
                             defaultPositionalArguments.get(variableIndex - defaultPositionalStartIndex));
                 } else {
-                    out = out.addPositionalOnlyArgument(co_varnames.get(variableIndex), PythonLikeObject.class);
+                    out = out.addPositionalOnlyArgument(co_varnames.get(variableIndex),
+                            (Class) parameterTypeList.get(variableIndex)
+                                    .getJavaClassOrDefault(PythonLikeObject.class));
                 }
             }
 
             for (; variableIndex < co_argcount; variableIndex++) {
                 if (variableIndex >= defaultPositionalStartIndex) {
-                    out = out.addArgument(co_varnames.get(variableIndex), PythonLikeObject.class,
+                    out = out.addArgument(co_varnames.get(variableIndex), (Class) parameterTypeList.get(variableIndex)
+                            .getJavaClassOrDefault(PythonLikeObject.class),
                             defaultPositionalArguments.get(variableIndex - defaultPositionalStartIndex));
                 } else {
-                    out = out.addArgument(co_varnames.get(variableIndex), PythonLikeObject.class);
+                    out = out.addArgument(co_varnames.get(variableIndex), (Class) parameterTypeList.get(variableIndex)
+                            .getJavaClassOrDefault(PythonLikeObject.class));
                 }
             }
 
             for (int i = 0; i < co_kwonlyargcount; i++) {
                 PythonLikeObject maybeDefault =
                         defaultKeywordArguments.get(PythonString.valueOf(co_varnames.get(variableIndex)));
                 if (maybeDefault != null) {
-                    out = out.addKeywordOnlyArgument(co_varnames.get(variableIndex), PythonLikeObject.class,
+                    out = out.addKeywordOnlyArgument(co_varnames.get(variableIndex),
+                            (Class) parameterTypeList.get(variableIndex)
+                                    .getJavaClassOrDefault(PythonLikeObject.class),
                             maybeDefault);
                 } else {
-                    out = out.addKeywordOnlyArgument(co_varnames.get(variableIndex), PythonLikeObject.class);
+                    out = out.addKeywordOnlyArgument(co_varnames.get(variableIndex),
+                            (Class) parameterTypeList.get(variableIndex)
+                                    .getJavaClassOrDefault(PythonLikeObject.class));
                 }
```

Comment:
This is considerable code duplication. Please consider extracting the shared code into a well-named method.